crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use aws_config::default_provider::credentials::DefaultCredentialsChain;
+use aws_config::BehaviorVersion;
+use aws_credential_types::provider::ProvideCredentials;
+use object_store::aws::AmazonS3Builder;
+use object_store::{MultipartUpload, ObjectStore, PutPayload};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use tokio::io::AsyncReadExt;
+use tracing::{debug, trace};
+use url::Url;
+
+use restate_partition_store::snapshots::{PartitionSnapshotMetadata, SnapshotFormatVersion};
+use restate_types::config::SnapshotsOptions;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::logs::Lsn;
+
+/// Provides read and write access to the long-term partition snapshot storage destination.
+///
+/// The repository wraps access to an object store "bucket" that contains snapshot metadata and data
+/// optimised for efficient retrieval. The bucket layout is split into two top-level prefixes for
+/// snapshot metadata and data respectively. While full snapshot archives contain all relevant
+/// metadata, this split layout allows for efficient retrieval of only the metadata upfront. It also
+/// enables us to evolve the data storage layout independently in the future.
+///
+/// A single top-level `latest.json` file is the only key which is repeatedly overwritten; all other
+/// data is immutable until the pruning policy allows for deletion.
+///
+/// - `[<prefix>/]<partition_id>/latest.json` - latest snapshot metadata for the partition
+/// - `[<prefix>/]<partition_id>/YYYY-MM-DD/{lsn}/metadata.json` - snapshot descriptor
+/// - `[<prefix>/]<partition_id>/YYYY-MM-DD/{lsn}/*.sst` - data files (explicitly named in `metadata.json`)
```

Comment:
Is the `YYYY-MM-DD` still up to date? I couldn't find it in the path when we upload snapshots.

Reply:
Oops, my bad! I kept flip-flopping on this but we can't have it half and half 😆 Taken out. We can make this configurable later if customers request it - either using dates or inverted padded sort-key for easier list navigation.